### PR TITLE
Throttle the time to call ServiceManagement for new rollout

### PR DESCRIPTION
**Pull request by qiwzhang**
https://github.com/cloudendpoints/esp/issues/651

### CODE REVIEW

File: src/api_manager/config_manager.cc [hunk 1]
```diff
@@ -20,86 +20,71 @@ namespace google {
...
 
-void ConfigManager::OnRolloutsRefreshTimer() {
-  global_context_->env()->LogDebug("RolloutsRefreshTimer start");
-  // If there is not any requests in the last window, not to call.
-  if (window_request_counter_ &&
-      window_request_counter_->Count(std::chrono::system_clock::now()) == 0) {
-    global_context_->env()->LogDebug(
-        "RolloutsRefresh skipped due to no traffic.");
+  // Timer is pending, or too close to last fetch time
+  if (fetch_timer_ || now < next_window_time_) {
```
Comment (JLXIA) [inline src/api_manager/config_manager.cc]: so logic decides whether to fetch new rollout or not, right?

should it be 

        if (!fetch_timer_ || now < next_window_time_) ?

Comment (qiwzhang) [reply src/api_manager/config_manager.cc]: No,  the opposite, do nothing.  If condition meets, do nothing, return.   if fetch_timer is still pending,  do nothing.  We only need one fetch timer.   when timer is fired,  fetch_timer will be null,  but it still have some time to fetch the service config.  so we set up next_window_time = timer_fired_time + window_size.    So we will not fire another timer before next window.

File: src/api_manager/config_manager.cc [hunk 2]
```diff
@@ -20,86 +20,71 @@ namespace google {
...
-    // Clear it to make sure it is fresh for the next timeout
-    global_context_->set_rollout_id("");
-    if (bail_out) {
-      skipped_rollout_calls_++;
-      global_context_->env()->LogDebug(
-          "RolloutsRefresh skipped due to response rollout_id.");
-      return;
-    }
-  }
+  auto throttled_time_in_ms = (*random_dist_)(random_generator_);
```
[the remark this answers is not in the document]
Comment (qiwzhang) [reply src/api_manager/config_manager.cc]: The whole idea is not for all ESP to call service_management at the same time.  This random delay is for such purpose.

File: src/api_manager/api_manager_impl.cc [hunk 3]
```diff
@@ -319,9 +323,6 @@ bool ApiManagerImpl::ReWriteURL(const char *uri, const size_t uri_len,
 
 std::unique_ptr<RequestHandlerInterface> ApiManagerImpl::CreateRequestHandler(
     std::unique_ptr<Request> request_data) {
-  if (config_manager_) {
-    config_manager_->CountRequests(1);
-  }
```
Comment (JLXIA) [inline src/api_manager/api_manager_impl.cc]: Why delete this?  Does that mean ESP still auto update, even no traffic?

Comment (qiwzhang) [reply src/api_manager/api_manager_impl.cc]: The old code is using every minute timer.  When a timer is fired,  use this one to check if there are any traffic,  if no traffic, do nothing.

The new code is not using periodical timer.  It purely uses the rollot_id from Check/Report response.   If rollot_id is changed,  fires a randomly throttled timer to fetch new service config.

If there is not traffic, will be no Check/Report response,  timer will not be fired.   So this code is not needed any more.

**Pull request merged**